### PR TITLE
Implement Array.prototype.filter method

### PR DESCRIPTION
**Pull request by Nickforall**
I implemented the `Array.prototype.filter` method according to the spec. Took the implementation of `map` as a reference, let me know if I missed something.

### CODE REVIEW

**Comment by Nickforall** [comment]
Rebased ✅

**Comment by Nickforall** [comment]
Rebased again due to conflicts from #268 😛

File: boa/src/builtins/array/tests.rs [hunk 1]
```diff
@@ -675,3 +675,69 @@ fn for_each_push_value() {
     assert_eq!(forward(&mut engine, "a[6]"), "6");
     assert_eq!(forward(&mut engine, "a[7]"), "8");
 }
+
```
Comment (jasonwilliams) [inline boa/src/builtins/array/tests.rs]: @Nickforall you don't have a #[test] attribute above the filter() function here like the others

Comment (Nickforall) [reply boa/src/builtins/array/tests.rs]: Whoops, must have happened while fixing merge conflicts

**Pull request merged**